File: problems/Average of Levels in Binary Tree/averageOfLevels.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<double> AverageOfLevels(TreeNode root) {
        var result = new List<double>();

        if (null == root) {
            return result;
        }

        var bfs = new Queue<TreeNode>();

        bfs.Enqueue(root);

        while (0 < bfs.Count) {
            var count = bfs.Count;
            double sum = 0;

            for (var i = 0; count > i; ++i) {
                var curr = bfs.Dequeue();

                sum += curr.val;
                if (null != curr.left) {
                    bfs.Enqueue(curr.left);
                }
                if (null != curr.right) {
                    bfs.Enqueue(curr.right);
                }
            }

            result.Add(sum / count);
        }

        return result;
    }
}

File: problems/Average of Levels in Binary Tree/averageOfLevels.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Double> averageOfLevels(TreeNode root) {
        List<Double> result = new ArrayList<>();
        Queue<TreeNode> bfs = new LinkedList<>();

        if (null != root) {
            bfs.add(root);
        }

        while (!bfs.isEmpty()) {
            double sum = 0;
            int count = bfs.size();

            for (int idx = bfs.size(); 0 < idx; --idx) {
                TreeNode node = bfs.poll();

                sum += node.val;
                if (null != node.left) {
                    bfs.add(node.left);
                }
                if (null != node.right) {
                    bfs.add(node.right);
                }
            }

            result.add(sum / count);
        }

        return result;
    }
}
